src/project/internal/recentprojectscontroller.cpp

```diff
@@ -0,0 +1,214 @@
+
+    return m_recentProjectsList;
+}
+
+Notification RecentProjectsController::recentProjectsListChanged() const
+{
+    return m_recentListChanged;
+}
+
+void RecentProjectsController::prependRecentFile(const RecentFile& newFile)
+{
+    if (newFile.empty()) {
+        return;
+    }
+
+    TRACEFUNC;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentProjectsList.size() + 1);
+    newList.push_back(newFile);
+
+    for (const RecentFile& file : m_recentProjectsList) {
+        if (file != newFile && fileSystem()->exists(file)) {
+            newList.push_back(file);
+        }
+    }
+
+    setRecentProjectsList(newList, true);
+
+    prependPlatformRecentFile(newFile);
+}
+
+void RecentProjectsController::clearRecentFiles()
+{
+    setRecentProjectsList({}, true);
+
+    clearPlatformRecentFiles();
+}
+
+void RecentProjectsController::prependPlatformRecentFile(const io::path_t&) {}
+
+void RecentProjectsController::clearPlatformRecentFiles() {}
+
+void RecentProjectsController::loadRecentProjectsList()
+{
+    ByteArray data = configuration()->rawRecentProjectsData();
+
+    RecentFilesList newList;
+
+    DEFER {
+        setRecentProjectsList(newList, false);
+    };
+
+    if (data.empty()) {
+        return;
+    }
+
+    QJsonParseError err;
+    QJsonDocument jsonDoc = QJsonDocument::fromJson(data.toQByteArrayNoCopy(), &err);
+    if (err.error != QJsonParseError::NoError) {
+        return;
+    }
+
+    if (!jsonDoc.isArray()) {
+        return;
+    }
+
+    for (const QJsonValue val : jsonDoc.array()) {
+        newList.push_back(val.toString());
+    }
+}
+
+void RecentProjectsController::removeNonexistentFiles()
+{
+    bool removed = false;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentProjectsList.size());
+
+    for (const RecentFile& file : m_recentProjectsList) {
+        if (fileSystem()->exists(file)) {
+            newList.push_back(file);
+        } else {
+            removed = true;
+        }
+    }
+
+    if (removed) {
+        setRecentProjectsList(newList, false);
+
+        async::Async::call(nullptr, [this, newList]() {
+            saveRecentProjectsList();
+        });
+    }
+}
+
+void RecentProjectsController::setRecentProjectsList(const RecentFilesList& list, bool save)
+{
+    m_recentProjectsList = list;
+
+    {
+        std::lock_guard lock(m_thumbnailCacheMutex);
+        m_thumbnailCache.clear();
+    }
+
+    if (save) {
+        saveRecentProjectsList();
+    }
+}
+
+void RecentProjectsController::saveRecentProjectsList()
+{
+    TRACEFUNC;
+
+    m_reloadingBlocked = true;
+
+    DEFER {
+        m_reloadingBlocked = false;
+    };
+
+    QJsonArray jsonArray;
+    for (const RecentFile& file : m_recentProjectsList) {
+        jsonArray << file.toQString();
+    }
+
+    QJsonDocument jsonDoc(jsonArray);
+    configuration()->setRawRecentProjectData(ByteArray::fromQByteArrayNoCopy(jsonDoc.toJson(QJsonDocument::Compact)));
```

Comment:
What do you think about saving this info to a separate json file instead of writing it to settings?

Reply:
That sounds like a good idea actually, I don't see why not. I'll try that out.